src/lem/circuit.rs

```diff
@@ -0,0 +1,904 @@
+    }
+}
+
+/// Manages global allocations for constants in a constraint system
+#[derive(Default)]
+pub(crate) struct GlobalAllocator<F: LurkField>(HashMap<FWrap<F>, AllocatedNum<F>>);
+
+#[inline]
+fn allocate_num<F: LurkField, CS: ConstraintSystem<F>>(
+    cs: &mut CS,
+    namespace: &str,
+    value: F,
+) -> Result<AllocatedNum<F>> {
+    AllocatedNum::alloc(cs.namespace(|| namespace), || Ok(value))
+        .with_context(|| format!("allocation for '{namespace}' failed"))
+}
+
+impl<F: LurkField> GlobalAllocator<F> {
+    /// Checks if the allocation for a numeric variable has already been cached.
+    /// If so, return the cached allocation variable. Allocate, cache and return
+    /// otherwise.
+    pub(crate) fn get_or_alloc_num<CS: ConstraintSystem<F>>(
+        &mut self,
+        cs: &mut CS,
+        f: F,
+    ) -> Result<AllocatedNum<F>> {
+        let wrap = FWrap(f);
+        match self.0.get(&wrap) {
+            Some(allocated_num) => Ok(allocated_num.to_owned()),
+            None => {
+                let allocated_num =
+                    allocate_num(cs, &format!("globally allocate {}", f.hex_digits()), f)?;
+                self.0.insert(wrap, allocated_num.clone());
+                Ok(allocated_num)
+            }
+        }
+    }
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, Hash)]
+pub(crate) enum SlotType {
+    Hash2,
+    Hash3,
+    Hash4,
+}
+
+impl SlotType {
+    pub(crate) fn preimg_size(&self) -> usize {
+        match self {
+            Self::Hash2 => 4,
+            Self::Hash3 => 6,
+            Self::Hash4 => 8,
+        }
+    }
+}
+
+impl std::fmt::Display for SlotType {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Hash2 => write!(f, "Hash2"),
+            Self::Hash3 => write!(f, "Hash3"),
+            Self::Hash4 => write!(f, "Hash4"),
+        }
+    }
+}
+
+/// A `Slot` is characterized by an index and a type
+#[derive(Clone, Copy, PartialEq, Eq, Hash)]
+struct Slot {
+    idx: usize,
+    typ: SlotType,
+}
+
+impl std::fmt::Display for Slot {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "Slot({}, {})", self.idx, self.typ)
+    }
+}
+
+impl LEM {
+    fn allocate_ptr<F: LurkField, CS: ConstraintSystem<F>>(
+        cs: &mut CS,
+        z_ptr: &ZPtr<F>,
+        name: &AString,
+        allocated_ptrs: &mut HashMap<AString, AllocatedPtr<F>>,
+    ) -> Result<AllocatedPtr<F>> {
+        if allocated_ptrs.contains_key(name) {
+            bail!("{} already allocated", name);
+        };
+        let allocated_tag =
+            allocate_num(cs, &format!("allocate {name}'s tag"), z_ptr.tag.to_field())?;
+        let allocated_hash = allocate_num(cs, &format!("allocate {name}'s hash"), z_ptr.hash)?;
+        let allocated_ptr = AllocatedPtr::from_parts(allocated_tag, allocated_hash);
+        allocated_ptrs.insert(name.clone(), allocated_ptr.clone());
+        Ok(allocated_ptr)
```

Comment:
Given that these all appear to be unconstrained allocations, is it definitely the case that all callers of `allocate_ptr` intend that? I'm trying to trace through and convince myself, but so far I'm not convinced. It would be useful to at least hear confirmation that you *intend* for all uses of `allocate_ptr` to be constrained  by their caller (or the caller's caller, or … at some other random place in the code…). My mysterious tone is because searching for the necessary constraints doesn't yet give me confidence. This is partially because if nothing else, the code is not written to make it *obvious* to me that allocations are all constrained. That would make verification easier and also increase my confidence even short of manually verification in each case.

Reply:
The tag can be a constant. 
We use `allocate_ptr` for input/output, which deserves a special discussion; and for other cases where an implication must appear later using the allocation, then it seems that some infrastructure may help to ensure this implications is indeed always present.

Reply:
So, looking at the instances of `allocate_ptr`, I see it being used in
1) `fn allocate_input`, which is okay, since the inputs do not necessarily have to be constrained (it will be in case it is used, as noted in a comment above)
2) `fn allocate_output`, which will be constrained by the returns, which should happen for every possible path, because of the very structure of LEM
3) `fn alloc_preimg` which is only used in the unhash macro, which constrains it by the `constrain_slot` macro
4) `allocated_img` in line 713, which is later constrained at line 731 and the `constrain_slot` macro
5) at line 782, in the Null case, where it is also constrained